### ISSUE TITLE
Unsure how to make structs act as a dict of values rather than a class containing variables that correspond to struct entries

### ISSUE THREAD

**Issue by Fhoughton** [swig/swig#2024]
I am attempting to create bindings to the C library [Raylib](https://github.com/raysan5/raylib) to use in combination with [Cython](https://github.com/cython/cython/). The issue comes from the fact that Cython expects structs to be python dictionaries when creating its static typing however when using the SWIG bindings I have made, which return an object which is a pointer to a class as is the default behaviour, an error occurs due to their incompatible types for the structs. I am aware of both wrappers and typemaps as potential solutions to the issue but am unsure how to implement them into my .i file. 

TL;DR: I need SWIG to make structs be a dictionary with key value pairs for each entry rather than a class with values for each struct entry

the error produced:
```cython
Traceback (most recent call last):
  File "main.pyx", line 44, in init main
    cdef raylib_structs.Texture2D texture = LoadTexture("resources/textures/texture.png")
  File "stringsource", line 14, in FromPyStructUtility.__pyx_convert__from_py_struct__Texture
TypeError: Expected a mapping, got Texture
```

raylib.i:
```
%module raylib
%{
#include "raylib-cpp.hpp"
#include "raylib.h"
%}

%include "raylib-cpp.hpp"
%include "raylib.h"
```

main.pyx (reduced):
```cython
import raylib
cimport raylib_structs

InitWindow(screenWidth, screenHeight, "raylib [core] example - basic window")
cdef raylib_structs.Texture2D texture = LoadTexture("resources/textures/texture.png")
SetTargetFPS(60)
CloseWindow()
```

raylib_structs.pxd (maps the structs to cython types, which become dicts with keys for each entry):
```cython
cdef extern from "raylib.h":
  cdef struct Texture:
      unsigned int id        

  ctypedef Texture Texture2D

  ctypedef Texture TextureCubemap
```

**Comment by ferdnyc** [comment]
@Fhoughton 

In Python, a type is basically defined by its interfaces. Provide the same interface methods, and you're a compatible type. So in order for your struct to be accessed as a `dict`, it needs to provide the `dict` methods. Not _all_ of them, but at least the core set. (But the more you implement, the closer to and more compatible with regular `dict` your struct will be.)

There's a section in the manual on "[Emulating container types](https://docs.python.org/3/reference/datamodel.html#emulating-container-types)" which gives the breakdown on what's required for each. For mappings (like `dict`), it advises:
> The first set of methods [referring to `__len__(self)` and `__length_hint__(self)`] is used either to emulate a sequence or to emulate a mapping; the difference is that for a sequence, the allowable keys should be the integers `k` for which `0 <= k < N` where `N` is the length of the sequence, or slice objects, which define a range of items. It is also recommended that mappings provide the methods `keys()`, `values()`, `items()`, `get()`, `clear()`, `setdefault()`, `pop()`, `popitem()`, `copy()`, and `update()` behaving similar to those for Python’s standard dictionary objects. The `collections.abc` module provides a `MutableMapping` abstract base class to help create those methods from a base set of `__getitem__()`, `__setitem__()`, `__delitem__()`, and `keys()`.
> [...]
> It is recommended that both mappings and sequences implement the `__contains__()` method to allow efficient use of the `in` operator; for mappings, `in` should search the mapping’s keys; for sequences, it should search through the values. It is further recommended that both mappings and sequences implement the `__iter__()` method to allow efficient iteration through the container; for mappings, `__iter__()` should iterate through the object’s keys; for sequences, it should iterate through the values.

You can choose to implement the methods either in C or in Python (well, it might be tricky in C due to the lack of support for real class methods) — for my C++ -> Python bindings, I chose a mix. I implemented the basics like `__len__()`, `__getitem__()`, `__contains__()`, and `__str__()` in C++, but from within the SWIG interface file (using `%extend ClassName { ... }`) so that the methods were only visible to the Python bindings, not the base C++ library. Then, I wrote the accessor methods like `__iter__()`, `keys()`, `items()`, and `values()` in Python, also in the interface file inside a `%pythoncode %{ ...%}` block.

Mixing things that way let me leverage C++'s direct access to the class members, for retrieving actual values, but then move to Python and benefit from its weak typing to avoid having to duplicate or overload methods in order to cover all of my disparate data types.

Mind you, by default the data members of the struct will _also_ be published in the bindings and accessible directly. With my `dict` methods implemented, the public members of my `Coordinate` class are still accessible from Python as `coord.x` and `coord.y`, but now I can also do this:
```python3
>>> from mylib import Coords
>>> from pprint import pp
>>> dict(Coords(1, 1))
{'x': 1, 'y': 1}
>>> points = [Coords(0, 0), Coords(4, 0), Coords(4, 4), Coords(0,4)]
>>> pp([{k: p[k] for k in p} for p in points])
[{'x': 0, 'y': 0},
 {'x': 0, 'y': 4},
 {'x': 4, 'y': 4},
 {'x': 4, 'y': 0}]
>>> # Or, more succinctly...
>>> pp([dict(q) for q in points])
[{'x': 0, 'y': 0},
 {'x': 0, 'y': 4},
 {'x': 4, 'y': 4},
 {'x': 4, 'y': 0}]
>>> # But, like I said, still...
>>> c = Coords(10, 20)
>>> c.x
10
>>> c.y
20
>>>
```

Personally I consider that a feature, not a bug, so I've never tried to change it. But if you really need to prevent direct access to the members, I'd bet there's probably some SWIG directive that you can use to make it hide the accessors it normally implements by default.

**Comment by ferdnyc** [comment]
Whoops, I realize I didn't demonstrate the _obvious_ simple usage. So, for the record, also:
```python3
>>> c["x"]
10
>>> c["y"]
20
>>>
```

**Comment by ferdnyc** [comment]
You don't need to overthink the method implementations, either. My Python implementation of `__iter__` is literally just:
```python3
def __iter__(self):
     return iter(self.items())
```
My `items()` is:
```python3
def items(self):
    _items = self._getDict()
    return _items.items()
```
Where `self._getDict()` is a class method that returns a `std::map<std::string, int>` — SWIG comes with include files that automatically generate implementations of things like Python's base type interfaces, for compatible STL classes.

**Comment by ojwb** [comment]
This seems to be a usage question and has a pretty detailed response so closing.
